File: bench/enterprise/java-spring/src/UserController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import java.sql.*;

@RestController
@RequestMapping("/api/users")
public class UserController {

    // VULNERABLE: SQL string concatenation of user input
    @GetMapping("/search")
    public String search(@RequestParam String q) throws Exception {
        Connection conn = DriverManager.getConnection("jdbc:postgresql://db/app");
        Statement st = conn.createStatement();
        ResultSet rs = st.executeQuery("SELECT * FROM users WHERE name LIKE '%" + q + "%'");
        return rs.getString(1);
    }

    // SAFE: PreparedStatement (parameterized)
    @GetMapping("/show")
    public String show(@RequestParam String id) throws Exception {
        Connection conn = DriverManager.getConnection("jdbc:postgresql://db/app");
        PreparedStatement ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        ps.setString(1, id);
        ResultSet rs = ps.executeQuery();
        return rs.getString(1);
    }

    // VULNERABLE: command execution from user input
    @GetMapping("/export")
    public String export(@RequestParam String file) throws Exception {
        Runtime.getRuntime().exec("tar -czf /tmp/export.tar.gz " + file);
        return "ok";
    }
}

File: bench/enterprise/php-laravel/app/Http/Controllers/UserController.php
<?php
namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use App\Models\User;

class UserController extends Controller
{
    // VULNERABLE: raw SQL interpolation of user input
    public function search(Request $request)
    {
        $q = $request->input('q');
        $users = DB::select("SELECT * FROM users WHERE name LIKE '%{$q}%'");
        return response()->json($users);
    }

    // VULNERABLE: raw where with user input (query builder bypass)
    public function filter(Request $request)
    {
        $col = $request->input('col');
        $rows = DB::table('users')->whereRaw("{$col} = 1")->get();
        return response()->json($rows);
    }

    // SAFE: Eloquent ORM (parameterized)
    public function show(Request $request)
    {
        $id = $request->input('id');
        $user = User::where('id', $id)->first();
        return response()->json($user);
    }
}
